subsys/usb/class/netusb/netusb.c

```diff
@@ -21,88 +21,69 @@
 #include "../../usb_descriptor.h"
 #include "../../composite.h"
 #include "netusb.h"
-#include "eth_emu.h"
-#include "function_ecm.h"
 
-#define NETUSB_NUM_CONF 1
+#define DEV_DATA(dev) ((struct netusb_dev * const)(dev)->driver_data)
+#define NETDEV_DATA(iface) DEV_DATA(net_if_get_device(iface))
 
-static struct __netusb {
+static struct netusb_dev {
+	struct usb_cfg_data usb_config;
+	struct netusb_function *func;
+	struct net_if *iface;
+	u8_t mac_addr[6];
 	u8_t conf;
-	u8_t conf_index;
-	struct netusb_function *func[NETUSB_NUM_CONF];
-} netusb;
-
-/* Ethernet Emulation device */
-static struct eth_emu_context *eth_emu_ctx;
+} netusb0;
 
-int netusb_send(struct net_pkt *pkt)
+static int netusb_send(struct net_if *iface, struct net_pkt *pkt)
 {
+	struct netusb_dev *net = NETDEV_DATA(iface);
+	int ret;
+
 	SYS_LOG_DBG("Send pkt, len %u", net_pkt_get_len(pkt));
 
-	switch (netusb.conf) {
-	case 1:
-		return ecm_send(pkt);
-	default:
-		SYS_LOG_ERR("Unconfigured device, conf %u", netusb.conf);
-		break;
+	if (!net->conf) {
+		SYS_LOG_ERR("Unconfigured device, conf");
+		return -ENODEV;
 	}
 
-	return -ENODEV;
-}
-
-static struct usb_ep_cfg_data netusb_ep_data[] = {
-	/* Configuration ECM */
-	{
-		.ep_cb = ecm_int_in,
-		.ep_addr = CONFIG_CDC_ECM_INT_EP_ADDR
-	},
-	{
-		.ep_cb = ecm_bulk_out,
-		.ep_addr = CONFIG_CDC_ECM_OUT_EP_ADDR
-	},
-	{
-		.ep_cb = ecm_bulk_in,
-		.ep_addr = CONFIG_CDC_ECM_IN_EP_ADDR
-	},
-};
+	if (!net->func) {
+		SYS_LOG_ERR("No netusb function selected");
+		return -ENODEV;
+	}
 
-static int netusb_connect_media(void)
-{
-	if (!netusb.func[netusb.conf_index]->connect_media) {
-		return -ENOTSUP;
+	ret = net->func->send_pkt(pkt);
+	if (ret) {
+		SYS_LOG_ERR("No netusb function selected");
+		return ret;
 	}
 
-	return netusb.func[netusb.conf_index]->connect_media(true);
+	net_pkt_unref(pkt);
+
+	return 0;
 }
 
-static int netusb_disconnect_media(void)
+int netusb_recv(struct net_pkt *pkt)
 {
-	if (!netusb.func[netusb.conf_index]->connect_media) {
-		return -ENOTSUP;
-	}
+	SYS_LOG_DBG("Recv pkt, len %u", net_pkt_get_len(pkt));
 
-	return netusb.func[netusb.conf_index]->connect_media(false);
+	return net_recv_data(netusb0.iface, pkt);
 }
 
 static void netusb_status_configured(u8_t *conf)
 {
 	SYS_LOG_INF("Enable configuration number %u", *conf);
 
-	netusb.conf = *conf;
+	netusb0.conf = *conf;
 
 	switch (*conf) {
 	case 1:
-		netusb.conf_index = 0;
-		eth_emu_up();
-		netusb_connect_media();
```

Comment:
connect_media() callback is needed for RNDIS and probably other protocol, is it needed for EEM? Are you developing EEM at the moment?

Reply:
yes I do, but wanted to avoid introducing unused cb, could come with RNDIS or EEM. Since this cb was already present I can probably keep it.